nixos/modules/services/web-apps/immich.nix

```diff
@@ -0,0 +1,574 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}: let
+  inherit
+    (lib)
+    hasAttr
+    hasPrefix
+    maintainers
+    mapAttrs
+    mkDefault
+    mkEnableOption
+    mkIf
+    mkMerge
+    mkOption
+    mkPackageOption
+    optional
+    optionalAttrs
+    optionalString
+    types
+    ;
+
+  cfg = config.services.immich;
+  serverCfg = config.services.immich.server;
+  backendCfg = serverCfg.backend;
+  microservicesCfg = serverCfg.microservices;
+  webCfg = cfg.web;
+  mlCfg = cfg.machineLearning;
+
+  isServerPostgresUnix = hasPrefix "/" serverCfg.postgres.host;
+  postgresEnv =
+    if isServerPostgresUnix
+    then {
+      # If passwordFile is given, this will be overwritten in ExecStart
+      DB_URL = "socket://${serverCfg.postgres.host}?dbname=${serverCfg.postgres.database}";
+    }
+    else {
+      DB_HOSTNAME = serverCfg.postgres.host;
+      DB_PORT = toString serverCfg.postgres.port;
+      DB_DATABASE_NAME = serverCfg.postgres.database;
+      DB_USERNAME = serverCfg.postgres.username;
+    };
+
+  typesenseEnv = {
+    TYPESENSE_ENABLED = toString serverCfg.typesense.enable;
+  } // optionalAttrs serverCfg.typesense.enable {
+    TYPESENSE_HOST = serverCfg.typesense.host;
+    TYPESENSE_PORT = toString serverCfg.typesense.port;
+    TYPESENSE_PROTOCOL = serverCfg.typesense.protocol;
+  };
+
+  # Don't start a redis instance if the user sets a custom redis connection
+  enableRedis = !hasAttr "REDIS_URL" serverCfg.extraConfig && !hasAttr "REDIS_SOCKET" serverCfg.extraConfig;
+  redisServerCfg = config.services.redis.servers.immich;
+  redisEnv = optionalAttrs enableRedis {
+    REDIS_SOCKET = redisServerCfg.unixSocket;
+  };
+
+  serverEnv = postgresEnv // typesenseEnv // redisEnv // {
+    NODE_ENV = "production";
+
+    IMMICH_MEDIA_LOCATION = serverCfg.mediaDir;
+    IMMICH_MACHINE_LEARNING_URL =
+      if serverCfg.machineLearningUrl != null
+      then serverCfg.machineLearningUrl
+      else "false";
+  };
+
+  serverStartWrapper = program: ''
+    set -euo pipefail
+    mkdir -p ${serverCfg.mediaDir}
+
+    ${optionalString (serverCfg.postgres.passwordFile != null) (
+      if isServerPostgresUnix
+      then ''export DB_URL="socket://${serverCfg.postgres.username}:$(cat ${serverCfg.postgres.passwordFile})@${serverCfg.postgres.host}?dbname=${serverCfg.postgres.database}"''
+      else "export DB_PASSWORD=$(cat ${serverCfg.postgres.passwordFile})"
+    )}
+
+    ${optionalString serverCfg.typesense.enable ''
+      export TYPESENSE_API_KEY=$(cat ${serverCfg.typesense.apiKeyFile})
+    ''}
```

Comment:
immich-server looks for `TYPESENSE_API_KEY` even if `TYPESENSE_ENABLE=false`, so when `cfg.server.typesense.enable = false`, immich-server won't start.